hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/CommitMetadataUtils.java

```diff
@@ -0,0 +1,220 @@
+                HoodieDeltaWriteStat originalWriteStat =
+                    (HoodieDeltaWriteStat) fileIdToOriginalWriteStat.get(fileId); // are there chances that there won't be any write stat in original list?
+                logFileStatuses.forEach(fileStatus -> {
+                  // for every missing file, add a new HoodieDeltaWriteStat
+                  HoodieDeltaWriteStat writeStat = new HoodieDeltaWriteStat();
+                  HoodieLogFile logFile = new HoodieLogFile(fileStatus);
+                  writeStat.setPath(basePathLocal, logFile.getPath());
+                  writeStat.setPartitionPath(partitionPath);
+                  writeStat.setFileId(fileId);
+                  writeStat.setTotalWriteBytes(logFile.getFileSize());
+                  writeStat.setFileSizeInBytes(logFile.getFileSize());
+                  writeStat.setLogVersion(logFile.getLogVersion());
+                  List<String> logFiles = new ArrayList<>(originalWriteStat.getLogFiles());
+                  logFiles.add(logFile.getFileName());
+                  writeStat.setLogFiles(logFiles);
+                  writeStat.setBaseFile(originalWriteStat.getBaseFile());
+                  writeStat.setPrevCommit(logFile.getBaseCommitTime());
+                  missingWriteStats.add(writeStat);
+                });
+              });
+              return Pair.of(partitionPath, missingWriteStats);
+            }).collectAsList();
+
+        // add these write stat to commit meta. deltaWriteStat can be empty due to file missing. See code above to address FileNotFoundException
+        for (Pair<String, List<HoodieWriteStat>> partitionDeltaStats : additionalLogFileWriteStat) {
+          String partitionPath = partitionDeltaStats.getKey();
+          partitionDeltaStats.getValue().forEach(ws -> commitMetadata.addWriteStat(partitionPath, ws));
+        }
+      }
+    }
+    return commitMetadata;
+  }
+
+  private static HoodiePairData<String, Map<String, HoodieWriteStat>> getPartitionToFileIdToFilesMap(HoodieCommitMetadata commitMetadata, HoodieEngineContext context) {
+
+    List<Map.Entry<String, List<HoodieWriteStat>>> partitionToWriteStats = new ArrayList(commitMetadata.getPartitionToWriteStats().entrySet());
+
+    return context.parallelize(partitionToWriteStats)
+        .mapToPair((SerializablePairFunction<Map.Entry<String, List<HoodieWriteStat>>, String, Map<String, HoodieWriteStat>>) t -> {
+          Map<String, HoodieWriteStat> fileIdToWriteStat = new HashMap<>();
+          t.getValue().forEach(writeStat -> {
+            if (!fileIdToWriteStat.containsKey(writeStat.getFileId())) {
+              fileIdToWriteStat.put(writeStat.getFileId(), writeStat);
+            }
+          });
+          return Pair.of(t.getKey(), fileIdToWriteStat);
+        });
+  }
+
+  private static HoodiePairData<String, Map<String, List<String>>> getPartitionToFileIdToMissingLogFileMap(String basePathStr, Set<String> logFilesMarkerPath, HoodieEngineContext context,
```

Comment:
this looks ok to me. 
after this, we have to do the join anyways so that we can create write stat for each missing file. 

instead, in the cur logic, we are keeping everything in one join and map call. 
lmk if you think we have compelling reason to go w/ your suggestion